Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs

```diff
@@ -0,0 +1,272 @@
+
+        protected override Task HandleAsync(
+            IKernelCommand command,
+            KernelInvocationContext context)
+        {
+            if (command is KernelCommandBase kb)
+            {
+                if (kb.Handler == null)
+                {
+                    switch (command)
+                    {
+                        case SubmitCode submitCode:
+                            submitCode.Handler = async (_, invocationContext) =>
+                            {
+                                await HandleSubmitCode(submitCode, context);
+                            };
+                            break;
+
+                        case RequestCompletion requestCompletion:
+                            requestCompletion.Handler = async (_, invocationContext) =>
+                            {
+                                await HandleRequestCompletion(requestCompletion, invocationContext);
+                            };
+                            break;
+
+                        case CancelCurrentCommand interruptExecution:
+                            interruptExecution.Handler = async (_, invocationContext) =>
+                            {
+                                await HandleCancelCurrentCommand(interruptExecution, invocationContext);
+                            };
+                            break;
+                    }
+                }
+            }
+
+            return Task.CompletedTask;
+        }
+
+        #endregion
+
+        #region Handlers
+
+        private async Task HandleSubmitCode(
+                SubmitCode submitCode,
+                KernelInvocationContext context)
+        {
+            CancellationTokenSource cancellationSource;
+            lock (_cancellationSourceLock)
+            {
+                cancellationSource = _cancellationSource;
+            }
+            
+            // Acknowledge that we received the request.
+            var codeSubmissionReceived = new CodeSubmissionReceived(submitCode);
+            context.Publish(codeSubmissionReceived);
+
+            // Test is the code we got is actually able to run.
+            var code = submitCode.Code;
+            if (IsCompleteSubmission(code))
+            {
+                context.Publish(new CompleteCodeSubmissionReceived(submitCode));
+            }
+            else
+            {
+                context.Publish(new IncompleteCodeSubmissionReceived(submitCode));
+            }
+
+            // Do nothing if we get a Diagnose type.
+            if (submitCode.SubmissionType == SubmissionType.Diagnose)
+            {
+                return;
+            }
+
+            if (cancellationSource.IsCancellationRequested)
+            {
+                context.Fail(null, "Command cancelled");
+                return;
+            }
+        
+            // Wait until we have access to the runspace since we only can run one cell at a time.
+            await _runspaceSemaphore.WaitAsync();
```

Comment:
Is it possible that multiple requests coming in parallel? A Jupyter client will spins up a kernel, but can a kernel be shared by multiple clients?

Reply:
Yeah you can trigger multiple cells at the same time so this protects the runspace from a second invocation trying to take the runspace.

Reply:
The command pipeline should prevent parallel execution. While the Jupyter client lets you send a new submission  while a previous one is executing, the kernel forces them to run serially.

Reply:
Great to know that (I guess that's done by the [Aggregattion here](https://github.com/dotnet/interactive/blob/master/Microsoft.DotNet.Interactive/KernelCommandPipeline.cs#L59-L63)).
Then we probably can remove the semaphore.

Reply:
What about the completion message? Would that also wait for the execution of code to finish before handling the completion request?

Reply:
The trampoline here is intended to enforce serial execution regardless of the kind of command sent to the kernel: https://github.com/dotnet/interactive/blob/a49941b40730cebc65dd73af6c65401e0a9fc06b/Microsoft.DotNet.Interactive/KernelBase.cs#L192-L202

If you see any commands executing in parallel, consider it a bug.

Reply:
I've tested this and they look like they're handled serially as well.

Reply:
I've removed the semaphore 👍